### PR TITLE
Navigation Service, Part 1

### PR DESCRIPTION
**Pull request by JThramer**
It's working!

Part 1 of the "Nav To The Future" initiative.

To-Do: 
- [x] Get tests working
- [x] refactor `EventsManager`
- [x] refactor `TunnelIntersectionManager` (now `TunnelAuthority`)
- [x] Fix weird CI bug where we're failing strings validation step
- [x]  refactor `TunnelIntersectionManagerTests`
- [x] PR Comments
- [x] refactor delegation protocols
- [ ] Update Docs
- [ ] Smoke Test
- [ ] Update `CHANGELOG.md`

/cc @mapbox/navigation-ios @mcwhittemore

### CODE REVIEW

File: MapboxCoreNavigation/NavigationService.swift [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    func stop()
+}
+
+@objc(MBNavigationService)
+public class MapboxNavigationService: NSObject, NavigationService {
+    public var locationSource: NavigationLocationManager
+    var directionsService: Directions
+    public var router: Router
+    public var eventsManager: EventsManager
+    public var delegate: NavigationServiceDelegate?
```
Comment (akitchen) [inline MapboxCoreNavigation/NavigationService.swift]: Should this be `weak`?

File: MapboxNavigation/NavigationViewController.swift [hunk 2]
```diff
@@ -213,18 +213,28 @@ open class NavigationViewController: UIViewController {
...
+            return navigationService.route
+        }
+        set {
+            navigationService.route = newValue
             NavigationSettings.shared.distanceUnit = route.routeOptions.locale.usesMetric ? .kilometer : .mile
-            mapViewController?.notifyDidReroute(route: route)
+                        mapViewController?.notifyDidReroute(route: route)
         }
     }
+//    @objc public var route: Route! {
```
Comment (akitchen) [inline MapboxNavigation/NavigationViewController.swift]: What's happening here? I'd rather not bring commented-out code into master

Comment (JThramer) [reply MapboxNavigation/NavigationViewController.swift]: Apologies, that shouldn't of made it into the PR. 😅

File: MapboxNavigation/NavigationViewController.swift [hunk 3]
```diff
@@ -602,16 +618,13 @@ extension NavigationViewController: RouteControllerDelegate {
         }
         return advancesToNextLeg
     }
-}
-
-extension NavigationViewController: TunnelIntersectionManagerDelegate {
     public func tunnelIntersectionManager(_ manager: TunnelIntersectionManager, willEnableAnimationAt location: CLLocation) {
-        routeController.tunnelIntersectionManager(manager, willEnableAnimationAt: location)
+//        routeController.tunnelIntersectionManager(manager, willEnableAnimationAt: location)
```
Comment (akitchen) [inline MapboxNavigation/NavigationViewController.swift]: This seems like a TODO

File: MapboxNavigation/NavigationViewController.swift [hunk 4]
```diff
@@ -602,16 +618,13 @@ extension NavigationViewController: RouteControllerDelegate {
...
         styleManager.applyStyle(type: .night)
     }
     
     public func tunnelIntersectionManager(_ manager: TunnelIntersectionManager, willDisableAnimationAt location: CLLocation) {
-        routeController.tunnelIntersectionManager(manager, willDisableAnimationAt: location)
+//        routeController.tunnelIntersectionManager(manager, willDisableAnimationAt: location)
```
Comment (akitchen) [inline MapboxNavigation/NavigationViewController.swift]: Another TODO ?

File: MapboxNavigation/NavigationViewController.swift [hunk 5]
```diff
@@ -349,17 +359,23 @@ open class NavigationViewController: UIViewController {
...
+        navigationService.delegate = self
         
-        self.directions = directions
-        self.route = route
-        NavigationSettings.shared.distanceUnit = route.routeOptions.locale.usesMetric ? .kilometer : .mile
-        routeController.resume()
+        navigationService.start()
         
-        let mapViewController = RouteMapViewController(routeController: self.routeController, delegate: self)
+//        self.routeController = RouteController(along: route, directions: directions, locationManager: locationManager ?? NavigationLocationManager())
```
[the remark this answers is not in the document]
Comment (JThramer) [reply MapboxNavigation/NavigationViewController.swift]: Ugh, this should of made it in either. Sourcetree is being difficult today.

**Comment by akitchen** [comment]
>  Get tests working
 Please focus here next. Refactoring without test coverage is 👎

File: MapboxCoreNavigation/NavigationService.swift [hunk 6]
```diff
@@ -0,0 +1,110 @@
+import Foundation
+import CoreLocation
+import MapboxDirections
+
+@objc public protocol NavigationServiceDelegate: RouterDelegate, TunnelIntersectionManagerDelegate {}
```
[the remark this answers is not in the document]
Comment (JThramer) [reply MapboxCoreNavigation/NavigationService.swift]: @1ec5 You're totally right here, and `NavigationService` is only composed as-is for purposes of convenience. I am planning on doing exactly that you suggested as part of the `refactor delegation protocols` task.

File: MapboxNavigation/NavigationViewController.swift [hunk 7]
```diff
@@ -213,16 +213,14 @@ open class NavigationViewController: UIViewController {
...
-            }
+    @objc public var route: Route {
+        get {
+            return navigationService.route
+        }
+        set {
+            navigationService.route = newValue
             NavigationSettings.shared.distanceUnit = route.routeOptions.locale.usesMetric ? .kilometer : .mile
-            mapViewController?.notifyDidReroute(route: route)
+                        mapViewController?.notifyDidReroute(route: route)
```
[the remark this answers is not in the document]
Comment (JThramer) [reply MapboxNavigation/NavigationViewController.swift]: - [ ] Fix this

File: MapboxCoreNavigationTests/MapboxCoreNavigationTests.swift [hunk 8]
```diff
@@ -130,20 +130,23 @@ class MapboxCoreNavigationTests: XCTestCase {
...
         }
         
-        expectation(forNotification: .routeControllerDidFailToReroute, object: navigation) { (notification) -> Bool in
+        expectation(forNotification: .routeControllerDidFailToReroute, object: navigation.router) { (notification) -> Bool in
             return true
         }
-        
-        navigation.reroute(from: CLLocation(latitude: 0, longitude: 0), along: navigation.routeProgress)
-        directionsClientSpy.fireLastCalculateCompletion(with: nil, routes: nil, error: NSError())
+        let routeController = navigation.router as! RouteController
```
Comment (JThramer) [inline MapboxCoreNavigationTests/MapboxCoreNavigationTests.swift]: If `reroute(from:along)` stays public, this forced-cast is unnecessary.

File: MapboxNavigationTests/NavigationViewControllerTests.swift [hunk 9]
```diff
@@ -156,19 +156,13 @@ class NavigationViewControllerTests: XCTestCase {
...
         UIApplication.shared.delegate!.window!!.addSubview(navigationViewController.view)
         
         let service = dependencies.navigationService
         
         // Identify a location without a custom road name.
         let fultonStreetLocation = dependencies.poi[2]
-//        let fultonStreetLocation = service.router.route.legs.first!.steps[5].intersections!.first!.location
+
         
         navigationViewController.mapViewController!.labelRoadNameCompletionHandler = { (defaultRaodNameAssigned) in
```
Comment (JThramer) [inline MapboxNavigationTests/NavigationViewControllerTests.swift]: `defaultRaodNameAssigned`

File: MapboxCoreNavigation/EventDetails.swift [hunk 10]
```diff
@@ -105,19 +104,15 @@ struct EventDetails: Encodable {
...
         rerouteCount = session.numberOfReroutes
         
-        if let manager = router.locationManager {
-            locationEngine = String(describing: manager)
-            locationManagerDesiredAccuracy = manager.desiredAccuracy
-        } else {
-            locationEngine = nil
-            locationManagerDesiredAccuracy = nil
-        }
+        locationEngine = dataSource.locationSource.description
```
Comment (frederoni) [inline MapboxCoreNavigation/EventDetails.swift]: Use the name of the class instead. `LocationSource.description` only supports simulation or device but you can also simulate on a device, or replay using the simulator, or use a custom location source. This information is valuable when debugging events.

File: MapboxCoreNavigation/NavigationLocationManager.swift [hunk 11]
```diff
@@ -15,6 +15,14 @@ import UIKit
 open class NavigationLocationManager: CLLocationManager {
     
     var lastKnownLocation: CLLocation?
+    override open var delegate: CLLocationManagerDelegate? {
```
Comment (frederoni) [inline MapboxCoreNavigation/NavigationLocationManager.swift]: nit: no need to override the default implementation for delegate with the same implementation as the superclass already provides.

File: MapboxCoreNavigation/NavigationService.swift [hunk 12]
```diff
@@ -0,0 +1,276 @@
+import Foundation
+import CoreLocation
+import MapboxDirections
+
+
+@objc public enum LocationSource: Int {
```
Comment (JThramer) [inline MapboxCoreNavigation/NavigationService.swift]: Unused.

File: MapboxCoreNavigation/EventDetails.swift [hunk 13]
```diff
@@ -306,8 +301,8 @@ extension EventDetails {
         }
     }
     
-    static func defaultEvents(router: Router) -> EventDetails {
-        return EventDetails(router: router, session: router.eventsManager.sessionState)
+    static func defaultEvents(dataSource source: EventsManagerDataSource, session state: SessionState) -> EventDetails {
```
Comment (JThramer) [inline MapboxCoreNavigation/EventDetails.swift]: This is redundant.

File: MapboxCoreNavigation/RouteController.swift [hunk 14]
```diff
@@ -163,36 +143,17 @@ open class RouteController: NSObject, Router {
...
-    }
-    
-    /**
-     Ends the current navigation session.
-     */
-    @objc public func endNavigation(feedback: EndOfRouteFeedback? = nil) {
-        suspendLocationUpdates()
-        suspendNotifications()
-    }
+//    /**
```
Comment (JThramer) [inline MapboxCoreNavigation/RouteController.swift]: Comments.

File: MapboxCoreNavigation/NavigationService.swift [hunk 15]
```diff
@@ -132,13 +160,20 @@ public class MapboxNavigationService: NSObject, NavigationService {
         eventsManager.resetSession()
         eventsManager.start()
     }
-    
-    private func enterTunnel() {
-        simulate(on: router.routeProgress)
+
+    private func resetGPSCountdown() {
+        // Immediately end simulation if it is occuring.
+        if simulatedLocationSource != nil, simulationMode == .onPoorGPS{
```
Comment (JThramer) [inline MapboxCoreNavigation/NavigationService.swift]: turn the simulationMode check into a guard. There's no reason to reset the timer if we're not on poor GPS mode.

File: MapboxCoreNavigation/CountdownTimer.swift [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    
+    private func scheduleTimer() {
+        timer.schedule(deadline: deadline, repeating: .never, leeway: accuracy)
+    }
+    
+    func arm() {
+        guard state == .disarmed else { return }
+        scheduleTimer()
+        timer.setEventHandler(handler: payload)
+        timer.resume()
```
Comment (JThramer) [inline MapboxCoreNavigation/CountdownTimer.swift]: Move state switch right under the check

File: MapboxCoreNavigation/CountdownTimer.swift [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+        timer.cancel()
+        scheduleTimer()
+        timer.resume()
+    }
+
+    func disarm() {
+        guard state == .armed else { return }
+        timer.cancel()
+        timer.setEventHandler {}
+        state = .disarmed
```
Comment (JThramer) [inline MapboxCoreNavigation/CountdownTimer.swift]: Move state change under check

File: MapboxCoreNavigation/NavigationService.swift [hunk 18]
```diff
@@ -55,33 +61,43 @@ public class MapboxNavigationService: NSObject, NavigationService {
...
+            simulate()
         }
     }
     
+    public static func isInTunnel(at location: CLLocation, along progress: RouteProgress) -> Bool {
+        return TunnelAuthority.isInTunnel(at: location, along: progress)
+    }
+
+    
     private func simulate(on potentialProgress: RouteProgress? = nil, intent: SimulationIntent = .manual) {
```
Comment (JThramer) [inline MapboxCoreNavigation/NavigationService.swift]: drop progress argument -- unneeded

**Pull request merged**